Show users with access to cash box feature (2402)

sfount opened the pull request:
This PR implements the feature requested in https://github.com/IMA-WorldHealth/bhima-2.X/issues/2402. 

It shows a list of users that have access to a given cash box on the Cash Box Management page, allowing quick review of which employees can use different cash windows. 

It also implements the component `bh-user`. This is a component that should be useful in a number of other places in the application, it shows a system user's display name, username, active status and time since they have last logged in. It can be used as follows: 

```html
<!-- Directly provide the user details object --> 
<bh-user details="user"></bh-user>

<!-- This component uses an experimental 'fetch' binding which -->
<!--  if found useful could be standardised. Including fetch in the --> 
<!-- component definition instructs the component to download the details -->
<bh-user id="5" fetch></bh-user>
```

Example of the `bh-user` component: 

![screenshot from 2018-09-01 23-38-28](https://user-images.githubusercontent.com/2844572/44950454-31b6be80-ae40-11e8-8a8c-332a1b48d6ce.png)

Example of cash box management, users with permission feature: 

![screenshot from 2018-09-01 23-22-39](https://user-images.githubusercontent.com/2844572/44950445-0207b680-ae40-11e8-86f0-4b6b04595cba.png)

Closes #2402

sfount committed:
feat(cashbox): server side list cashbox users

This commit implements the server side API and integration tests for
displaying all users with permission to access a specified cashbox.

GET /cashboxes/:id/users

sfount committed:
feat(cashbox): bh-cashbox-users component

This commit introduces the bh-cashbox-users component on the Cashbox
Management page. It also updates the clients cashbox service to expose
the /cashbox/:id/users route.

sfount committed:
feat(cashbox): bh-user component introduced

This commit introduces the `bh-filter` component for displaying an
individual system user.

This is the first component to test a `fetch` architecture that allows
it to either use a details object provided or to optionally take
responsability for fetching that data.

Example usage:
<bh-filter details="user"></bh-filter>
<bh-filter id="5" fetch></bh-filter>

This commit also updates the User service to the latest code layout
standardsd.

sfount committed:
chore(cashbox): udpate controller code for lint

This commit updates the cashboxes udpate controller to meet the linting
standards.

jniles commented on client/src/i18n/en/cashbox.json:
```diff
@@ -9,6 +9,17 @@
+    "USERS": { 
+      "TITLE": "Cashbox Users",
+      "SUBTITLE": "Users",
+      "INFO_1": "To update users with access to this cash box, use the",
```
Just FYI, it's possible to include HTML tags in these translation values, as well as variable substitutions.  An example of that is [in patient records](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/client/src/i18n/en/patient_records.json#L43).  I'm pretty sure you could even include a `ui-sref` with the addition of a [translate-compile](https://angular-translate.github.io/docs/#/guide/05_using-translate-directive#using-translate-directive_post-compiling) tag.

Personally, I prefer this method (even though it seems impure) since it allows you to preserve the context of the statement and allow developers to infer from the string what is going to happen.  It would look something like this for your code:
```js
"INFO" : "To update users with access to this cash box, use the <a href ui-sref="users.list">Users Management</a> link."
```

What do you think?

jniles replied:
Of course you'll need to then have `translate-compile` on the `translate` directive tag in your HTML.

jniles replied:
I would also add "The following users have access to this cashbox:" at the very end, just to be explicit.

sfount replied:
This is good to know about! I will follow this suggestion for this example; I think the context of what is being translated is lost slightly when broken into 3 pieces. 

In general though I think that it's concerning to put client application logic in i18n translation files. In this case it's fairly simple, I think any use of `<b>` or `<u>` or general style changes should be fine. Potentially we should be more careful when it comes to including elements that change the way the application works.

```html
<span translate translate-compile translate-sanitize-strategy="'sce'">
    CASHBOX.USERS.INFO_LINK
</span> 
``` 

Does read clearer than the previous implementation, however.

jniles commented on client/src/js/components/bhUser.js:
```diff
@@ -0,0 +1,59 @@
+ * @component bhUser
+ *
+ * @description
+ * Displays a sysyem users details.
```
This should probably be `system` :D

jniles commented on client/src/js/components/bhUser.js:
```diff
@@ -0,0 +1,59 @@
+ * responsible for getting the latest data.
+ *
+ * @example
+ * <bh-user details="user"></bh-user>
```
Nice examples!

jniles commented on test/integration/cashboxes.js:
```diff
@@ -154,6 +154,21 @@ describe('(/cashboxes) The Cashboxes API endpoint', () => {
       .catch(helpers.handler);
   });
 
+  it('GET /cashboxes/:id/users should return users subscribed to a cashbox', () => {
```
:dancer

jniles commented on server/controllers/admin/users/cashboxes.js:
```diff
@@ -26,10 +26,10 @@ function list(req, res, next) {
   db.exec(sql, [req.params.id])
-    .then((rows) => {
-      rows = rows.map(row => row.cashbox_id);
+    .then((cashboxes) => {
```
Helpful!

jniles reviewed (commented):
Code in general looks good to me.  Appreciate the lint updates!

My comment about angular-translate is at your discretion.  Personally, I think it is easier to reason about, but have never had feedback on it.  Let me know your thoughts.

jniles committed:
chore(i18n): add FR translations

This commit adds FR translations to the cashbox management module's user
interface.
